### PR TITLE
Adjust RecordDataFormatter to accept driver with invoke function

### PR DESCRIPTION
**Pull request by LuomaJuha**
Was lookin at the helper and it is used usually with only certain driver. Decided that if it should work like the record helper.

### CODE REVIEW

**Comment by demiankatz** [comment]
Thanks, @LuomaJuha, I think this is certainly an interesting idea and probably a step in the right direction, but I think we need to be careful about compatibility breaks here. A couple of initial thoughts:

1.) What if we adjusted the public methods so that if the first parameter is a record driver, they automatically set the driver and shift the other parameters left? This would allow old-style calls to continue working while also allowing the new-and-improved syntax. Perhaps there's an argument for waiting until we raise the minimum PHP version to 8 so that we can use union types in the typehints for these situations.

2.) The refactoring of the core metadata to a separate template isn't directly related to adding functionality to the helper. Maybe we should address part that as a separate PR, especially since that change can likely be merged more quickly and with less discussion.

**Comment by LuomaJuha** [comment]
@demiankatz bc was something I was wondering, but I will add functions for them also. Also template will be moved into a new pr. :)

**Comment by LuomaJuha** [comment]
@demiankatz Here is the new version with bc. There was only 1 function getData which needed to be taken into account.

File: module/VuFind/src/VuFind/View/Helper/Root/RecordDataFormatter.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/RecordDataFormatter.php]: I'd suggest `if ($args[0] instanceof RecordDriver) {` instead -- using the class directly instead of a string is better for static analysis.

File: module/VuFind/src/VuFind/View/Helper/Root/RecordDataFormatter.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/RecordDataFormatter.php:178]: Maybe we should also check that $this->driver is non-null at this point and throw an exception if it is, to avoid unexpected behavior caused by mixed legacy/non-legacy invocations.

File: module/VuFind/src/VuFind/View/Helper/Root/RecordDataFormatter.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/RecordDataFormatter.php]: I think you can just say RecordDriver in this context, for consistency with the use statement above.

File: module/VuFind/src/VuFind/View/Helper/Root/RecordDataFormatter.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/RecordDataFormatter.php]: I'd make "and or" "and/or" here.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @LuomaJuha, this is looking very good. Just a few minor suggestions; once those are addressed, I'll give it a more thorough test, and hopefully merge it!

File: module/VuFind/src/VuFind/View/Helper/Root/RecordDataFormatter.php
Comment (LuomaJuha) [reply module/VuFind/src/VuFind/View/Helper/Root/RecordDataFormatter.php:178]: Oh ye, I added a null check for this->driver a little bit higher in this part as a separate check.

File: module/VuFind/src/VuFind/View/Helper/Root/RecordDataFormatter.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/RecordDataFormatter.php]: We should probably initialize this to null, in case somebody finds a way to call getData without first calling __invoke (e.g. by pulling the object of the plugin manager).

File: module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/RecordDataFormatterTest.php
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/RecordDataFormatterTest.php]: In order to exercise both forms of usage, maybe we can use a `@dataProvider` that passes in the name of a method that receives a driver and spec and returns the data. Then we can run the test twice, once for legacy invocation and once for new invocation.

If you're not sure what I'm talking about, I can probably find some time next week to put together a clearer example (I'm off for the rest of this week).

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @LuomaJuha, nearly there, though I had a couple more minor thoughts!

File: module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/RecordDataFormatterTest.php
Comment (LuomaJuha) [reply module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/RecordDataFormatterTest.php]: Yeah, been doing something with the tests so I can try out and see it comes close to your thoughts!

File: module/VuFind/src/VuFind/View/Helper/Root/RecordDataFormatter.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/RecordDataFormatter.php:166]: Should we do the instanceof check BEFORE the empty check? Otherwise, if you invoke the method with a driver and an empty array, we'll end up doing extra work.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @LuomaJuha, this looks really good -- I have just one very small remaining suggestion. Once that's addressed, I think this can be merged, unless you plan to do any more work in this PR. If you want to adjust templates to use the new invocation, I wouldn't mind doing that here... but it could also be handled in a follow-up PR if you prefer, along with the other changes that we already discussed handling separately.

File: module/VuFind/src/VuFind/View/Helper/Root/RecordDataFormatter.php
Comment (LuomaJuha) [reply module/VuFind/src/VuFind/View/Helper/Root/RecordDataFormatter.php:166]: I could do that! Yeah, I can alter the templates to use the new invocation.

**Review by demiankatz — APPROVED** [review]
Thanks, @LuomaJuha, this is looking great, and the full integration test suite is passing, so I'm going to merge now. I'll look forward to the follow-up with further template refactoring whenever time permits. :-)

**Pull request merged**
